Add basic auth support to remote_calendar

Molier opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change

No breaking changes.

## Proposed change

Adds HTTP Basic Authentication support to the Remote Calendar integration.

If a calendar URL returns 401 with a `WWW-Authenticate: Basic` header, the config flow prompts for credentials in a second step. Credentials are stored in the config entry and used for ongoing fetches.

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue:
- This PR is related to issue:
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/43965
- Link to developer documentation pull request:
- Link to frontend pull request:

## Checklist

- [x] I understand the code I am submitting and can explain how it works.
- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

Thank you for contributing <3

allenporter reviewed (changes requested):
Hello, thank you for your contribution.

Please review this PR and revert all unnecessary changes that are not related to the functionality provided in the PR description (e.g. edits made by AI to `tests/common.py` and other logging or style changes).  If you would like to make style cleanups to existing code, then  send them independently. The reason to do this is to make it easier for reviewers to focus on new/different functionality from style cleanups.

allenporter commented on homeassistant/components/remote_calendar/config_flow.py:
```diff
@@ -47,26 +60,57 @@ async def async_step_user(
+                    dict(res.headers),
+                )
+                # Store URL and calendar name for auth step
+                self.url = user_input[CONF_URL]  # pylint: disable=attribute-defined-outside-init
```
Lets store `self.data = user_input` rather than the individual pieces, given we now have logic split across multiple steps it will be harder to manage.

allenporter commented on homeassistant/components/remote_calendar/client.py:
```diff
@@ -1,12 +1,74 @@
+        _LOGGER.debug("HTTP Basic Auth enabled for username: %s", username)
+
+    # Extract API key from URL query parameters and add as header
+    # Some services (like Radarr) prefer header-based authentication
```
Not sure i follow the rationale for this block. Taking a query parameter and also adding a header? Why is it a query parameter if the service prefers it be a header? This does not seem appropriate.

allenporter commented on homeassistant/components/remote_calendar/config_flow.py:
```diff
@@ -79,6 +123,98 @@ async def async_step_user(
+            try:
+                await parse_calendar(self.hass, res.text)
+            except InvalidIcsException:
+                errors["base"] = "invalid_ics_file"
```
Similarly here, sending the user back to the username and password form here won't work really right?

Also, don't love that this is duplicated across the steps.

Molier commented:
> Hello, thank you for your contribution.
> 
> Please review this PR and revert all unnecessary changes that are not related to the functionality provided in the PR description (e.g. edits made by AI to `tests/common.py` and other logging or style changes). If you would like to make style cleanups to existing code, then send them independently. The reason to do this is to make it easier for reviewers to focus on new/different functionality from style cleanups.

I've reverted the tests/common.py changes. 
However, the tests fail without this fix due to a voluptuous API change in Python 3.14 - vol.Schema objects are no longer directly iterable, requiring schema.schema instead. Should I open a separate PR for this fix, or is there another approach you'd prefer?

allenporter commented on homeassistant/components/remote_calendar/strings.json:
```diff
@@ -1,15 +1,27 @@
       "invalid_ics_file": "There was a problem reading the calendar information. See the error log for additional details.",
-      "timeout_connect": "[%key:common::config_flow::error::timeout_connect%]"
+      "timeout_connect": "[%key:common::config_flow::error::timeout_connect%]",
+      "unauthorized": "Authentication failed. Please verify your API key or credentials are correct."
```
How about we use `invalid_auth` as the error code and reuse `[%key:common::config_flow::error::invalid_auth]`?

Molier commented on homeassistant/components/remote_calendar/strings.json:
```diff
@@ -1,15 +1,27 @@
+          "username": "Username"
+        },
+        "data_description": {
+          "password": "Password for HTTP Basic Authentication (required by your reverse proxy or server).",
```
thats what the main use case was for me atleast? I have an awful lot of docker services running on a server that is served via reverse proxy(caddy) and i wanted the calenders to be available for home assistant,but reverse proxy everything is secured. And i imagine others also will encounter this issue when they are running stuff behind a reverse proxy. but indeed its perhaps too verbose/specific

Molier commented on homeassistant/components/remote_calendar/config_flow.py:
```diff
@@ -25,12 +25,67 @@
+
+        return (None, res)
+
     async def async_step_user(
```
thanks!

allenporter commented:
> when i pull in everything and run all the home assistant tasks so i have a clean dev env to work with, the tests fail. perhaps my env is pulling in a too new version of voluptuos, seems weird since i work in the dev container so everything should be perfectly as set by home assistant config? let me check what version its running as. but yes , pull this in without changing common and tests dont work
> 
> _edit_: it shows this for my 'uv tree':
> 
> ```
> ├── voluptuous v0.15.2
> ├── voluptuous-openapi v0.2.0
> │   └── voluptuous v0.15.2
> ├── voluptuous-serialize v2.7.0
> │   └── voluptuous v0.15.2
> ```

I'm not seeing any tests failing on the continuous integration. what's the failure?

Molier committed:
Add basic auth support to remote_calendar

When a calendar URL returns 401 with WWW-Authenticate: Basic,
prompt user for credentials in a second config flow step.
Credentials are stored in the config entry and passed through
the coordinator for ongoing calendar fetches.

allenporter reviewed (changes requested):
This looks pretty close, looking forward to merging. I just have some style suggestions that are hopefully straight forward to address.

Thank you for your work on this feature!

allenporter reviewed (approved):
Actually I think i can see all the changes reflected, i won't wait for your reply (just consider for next time).

I realize now there are some exception cleanups we can do but I'll send in a follow up.

allenporter reviewed (changes requested):
Can you fix the failing CI errors and mark ready for review when they are passing?

Molier committed:
Fix alphabetical key ordering in strings.json

Molier committed:
Apply ruff format and fix multi-exception catch for prek CI

allenporter reviewed (commented):
Still seeing formatting failures on CI.

Molier committed:
Revert ruff preview formatting to match CI ruff-format

Molier committed:
Use except (HTTPError, InvalidURL) as err: to match codebase style

allenporter reviewed (approved):
Please teach your claw to (a) resolve or reply to comments that are addressed and (b) use "ready for review" next time so reviewers don't have to guess when you're ready for review.  Thanks.
